src/Chapter3.hs

```diff
@@ -480,6 +522,54 @@ After defining the city, implement the following functions:
    and at least 10 living __people__ inside in all houses of the city totally.
 -}
 
+type Wall = Bool
+
+type Church = Bool
+type Library = Bool
+data Building = Library | Church
+
+data Castle = Castle
+    {
+      __name__ :: String,
+      wall :: Wall
+    }
+
+newtype Houses = Houses
+    {
+      __people__ :: Int
+    }
+
+data MagicalCity = MagicalCity
+    {
+      castle :: Castle,
+      building :: Building,
+      houses :: [Houses]
+    }
+
+
+buildCastle :: String -> MagicalCity -> MagicalCity
+buildCastle newname city = city { castle = newCastle }
+  where
+    ccity = castle city
+    newCastle = ccity { __name__ = newname }
+
+
+buildHouse :: Int -> MagicalCity -> MagicalCity
+buildHouse np city = city { houses = newHouse:hcity }
+  where
+    hcity = houses city
+    newHouse = Houses { __people__ = np }
```

Comment:
Interesting approach. It is usually called "smart constructor". However, to avoid building houses with more than 4 people you need to check it here, and also restrict usage of the Constructor and allow to create `House`s only with your `buildHouse` function.
You can read a bit more on the pattern in here: https://kowainik.github.io/posts/haskell-mini-patterns#smart-constructor

Generally, in here, it is possible and easy restrict that on type level, by using something else than `Int` 🙂